File: engine/notes/notes.go
package notes

import (
	"os"
	"path/filepath"
	"strings"
)

type Note struct {
	ID        string
	Name      string
	FilePath  string
	Namespace string
}

func DefaultNotesDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "state", "hornet", "notes")
}

func LegacyDbeeNotesDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "state", "nvim", "dbee", "notes")
}

// ListGlobalNotes returns all .sql files in ~/.local/state/hornet/notes/global/
// falling back to legacy ~/.local/state/nvim/dbee/notes/global/
func ListGlobalNotes() ([]Note, error) {
	dir := filepath.Join(DefaultNotesDir(), "global")
	_ = os.MkdirAll(dir, 0755)

	seen := make(map[string]bool)
	var results []Note

	// 1. Read from Hornet notes dir
	if entries, err := os.ReadDir(dir); err == nil {
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".sql") {
				name := strings.TrimSuffix(e.Name(), ".sql")
				seen[name] = true
				results = append(results, Note{
					ID:        e.Name(),
					Name:      name,
					FilePath:  filepath.Join(dir, e.Name()),
					Namespace: "global",
				})
			}
		}
	}

	// 2. Read from legacy nvim-dbee notes dir if present
	legacyDir := filepath.Join(LegacyDbeeNotesDir(), "global")
	if entries, err := os.ReadDir(legacyDir); err == nil {
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".sql") {
				name := strings.TrimSuffix(e.Name(), ".sql")
				if !seen[name] {
					seen[name] = true
					results = append(results, Note{
						ID:        e.Name(),
						Name:      name,
						FilePath:  filepath.Join(legacyDir, e.Name()),
						Namespace: "global",
					})
				}
			}
		}
	}

	return results, nil
}

func ReadNote(filePath string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func SaveNote(name, content string) (string, error) {
	if !strings.HasSuffix(name, ".sql") {
		name += ".sql"
	}
	dir := filepath.Join(DefaultNotesDir(), "global")
	_ = os.MkdirAll(dir, 0755)

	filePath := filepath.Join(dir, name)
	if err := os.WriteFile(filePath, []byte(content), 0644); err != nil {
		return "", err
	}
	return filePath, nil
}

File: src/notes.rs
use crate::models::NoteItem;
use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;

pub fn default_notes_dir() -> PathBuf {
    dirs::state_dir()
        .or_else(dirs::data_local_dir)
        .unwrap_or_else(|| PathBuf::from("."))
        .join("hornet")
        .join("notes")
}

pub fn legacy_notes_dir() -> PathBuf {
    dirs::state_dir()
        .or_else(dirs::data_local_dir)
        .unwrap_or_else(|| PathBuf::from("."))
        .join("nvim")
        .join("dbee")
        .join("notes")
}

pub fn list_notes() -> Result<Vec<NoteItem>, String> {
    let mut seen = HashSet::new();
    let mut results = Vec::new();

    let primary_dir = default_notes_dir().join("global");
    let _ = fs::create_dir_all(&primary_dir);

    // 1. Read Hornet notes
    if let Ok(entries) = fs::read_dir(&primary_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() && path.extension().map_or(false, |ext| ext == "sql") {
                if let Some(file_name) = path.file_stem().and_then(|s| s.to_str()) {
                    seen.insert(file_name.to_string());
                    results.push(NoteItem {
                        id: file_name.to_string(),
                        name: file_name.to_string(),
                        file_path: path.to_string_lossy().to_string(),
                        namespace: "global".to_string(),
                    });
                }
            }
        }
    }

    // 2. Read legacy nvim-dbee notes
    let legacy_dir = legacy_notes_dir().join("global");
    if let Ok(entries) = fs::read_dir(&legacy_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() && path.extension().map_or(false, |ext| ext == "sql") {
                if let Some(file_name) = path.file_stem().and_then(|s| s.to_str()) {
                    if seen.insert(file_name.to_string()) {
                        results.push(NoteItem {
                            id: file_name.to_string(),
                            name: file_name.to_string(),
                            file_path: path.to_string_lossy().to_string(),
                            namespace: "global".to_string(),
                        });
                    }
                }
            }
        }
    }

    Ok(results)
}

pub fn read_note(path: &str) -> Result<String, String> {
    fs::read_to_string(path).map_err(|e| format!("Failed to read note at {}: {}", path, e))
}

pub fn save_note(name: &str, content: &str) -> Result<String, String> {
    let mut file_name = name.to_string();
    if !file_name.ends_with(".sql") {
        file_name.push_str(".sql");
    }

    let dir = default_notes_dir().join("global");
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;

    let file_path = dir.join(&file_name);
    fs::write(&file_path, content).map_err(|e| e.to_string())?;

    Ok(file_path.to_string_lossy().to_string())
}
